multi-agent/lib/mas/resources/builtin_service.py

```
@@ -0,0 +1,773 @@
+            )
+            for dep_rid, orig_descriptor in reversed(originals):
+                try:
+                    if orig_descriptor is None:
+                        self._descriptor_repo.delete(dep_rid)
+                    else:
+                        self._descriptor_repo.save(orig_descriptor)
+                except Exception:
+                    logger.exception(
+                        "Failed to roll back promoted dependency '%s' — it may "
+                        "be left as PUBLIC and require manual cleanup", dep_rid,
+                    )
```

Ruff (0.16.1)

[warning] `try`-`except` within a loop incurs performance overhead

(PERF203)